docs/src/variables.md

```diff
@@ -580,6 +587,67 @@ julia> x = @variable(model, [i=1:2], base_name="x", lower_bound=i, integer=true)
     Creating two named JuMP variables with the same name results in an error at
     runtime. Use anonymous variables as an alternative.
 
+## Variables constrained on creation
+
+By default, `@variable(model, x)` creates a _free_ variable that belongs to the
+set of real numbers.
+
+If you add constraints, such as `@variable(model, x >= 0, Int)`, then the
+`@variable` macro is equivalent to:
+```julia
+@variable(model, x)
+@constraint(model, x in MOI.GreaterThan(0.0))
+@constraint(model, x in MOI.Integer())
+```
+Importantly, the bound and integrality constraints are added _after_ the
+variable has been created.
+
+However, in some cases, it is necessary to supply a constraining set _at
+creation time_. We call these variables _variables constrained on creation_ to
+differentiate them from _variables free on creation_.
+
+For example, one way to create a vector of variables constrained to the second
+order cone is as follows:
+```jldoctest constrained_variables; setup=:(model=Model())
+julia> @variable(model, x[1:3])
+3-element Array{VariableRef,1}:
+ x[1]
+ x[2]
+ x[3]
+
+julia> @constraint(model, x in SecondOrderCone())
+[x[1], x[2], x[3]] ∈ MathOptInterface.SecondOrderCone(3)
+```
+
+An alternative approach, creating a vector of variables constrained on creation
+to belong to the [`SecondOrderCone`](@ref) is:
+```jldoctest constrained_variables
+julia> @variable(model, y[1:3] in SecondOrderCone())
+3-element Array{VariableRef,1}:
+ y[1]
+ y[2]
+ y[3]
+```
+Importantly, in [Direct mode](@ref) and for some solver, the second-order
+constraint cannot be deleted without deleting the variables `y`.
+
+### The `set` keyword
+
+An alternate syntax to `x in Set` is to use the `set` keyword of
+[`@variable`](@ref). This is most useful when creating anonymous variables:
+```julia
+x = @variable(model, [1:2, 1:2], set = PSDCone())
+```
+
+!!! info
+    When using JuMP in [Direct mode](@ref), it may be required to constrain
```

Comment:
Perhaps note at the beginning of this section that the whole discussion is relevant only for direct mode, so that readers can skip it if they don't care about direct mode.